### ISSUE TITLE
EGL not found

### ISSUE THREAD

**Issue by motters** [GuillaumeLazar/qtrb#1]
Hi Pal,

When i run 30_configure_qt5_rpi.sh for some reason Qt is not detecting EGL. Can you think why?

```
 EGL .................................... no
  OpenVG ................................. no
  OpenGL:
    Desktop OpenGL ....................... no
    OpenGL ES 2.0 ........................ no
    OpenGL ES 3.0 ........................ no
    OpenGL ES 3.1 ........................ no
  Session Management ..................... yes
Features used by QPA backends:
  evdev .................................. yes
  libinput ............................... no
  INTEGRITY HID .......................... no
  mtdev .................................. no
  tslib .................................. no
  xkbcommon-evdev ........................ no
QPA backends:
  DirectFB ............................... no
  EGLFS .................................. no
```

I also had to add the below line to 20_mount_sysroot.sh below the mount statements to stop the "Illegal instruction" error 

```
sudo sed -i '/./s/^/#/g' $SYSROOT/etc/ld.so.preload
```

Really like the project idea, once working i was hoping to create a docker and vagrant file so it can be easily ran :D.

Is there any chance you could upload the created image + libs?

Sam

**Comment by GuillaumeLazar** [comment]
Hi Sam,

Thanks for your message. Qt rocks on RPi !

**The EGL / configure issue:**
What is your Raspbian version? Have you start your Raspberry Pi with the image before using my scripts? Some important processes are done on the first Raspberry Pi boot. 

The starting workflow is not yet explained, here are my steps: 
1. download 2017-09-07-raspbian-stretch and copy it on a sdcard
2. boot up the Raspberry pi with this sdcard
- expand file system
- enable SSH
- do the apt-get update / rpi-update
- shutdown the RPi and eject the sdcard
3. on your host, backup the sdcard to a sysroot .img file (qtrb_v0.1.img)
4. then you can use the  scripts concerning mount/configure/build etc

Ok for the "Illegal instruction error" fix, I never got this issue, maybe you will not get this error anymore with the workflow above

**The image + libs release:**
Yes I uploaded a QTRB v0.1, with all you need (target/device bin, host bin and sysroot). The sysroot image file must be used as the RaspberryPi OS on the sdcard **AND** to crosscompile your Qt project on your host). Check the [releases](https://github.com/GuillaumeLazar/qtrb/releases) page!

**Comment by motters** [comment]
Thanks for the quick reply,

I'll try the above and get back to you. 

The "2017-09-07-raspbian-stretch" is that the full raspbian image (with desktop) or the lite? I wouldn't mind making an image for the lite version to reduce overhead etc. 

Sam

**Comment by GuillaumeLazar** [comment]
You can see the Raspbian URL in the file [config.sh](https://github.com/GuillaumeLazar/qtrb/blob/master/config.sh), it's the full desktop version (not the lite).

QTRB support both X11 window application (XCB backend) or fullscreen OpenGL application (EGLFS backend). 

If you just want to reduce the overhead and don't need X11, you can disable the X11 auto start from `raspi-config`. If you also want to reduce the sysroot size, you could try raspbian lite but I never tried it.

**Comment by motters** [comment]
Sorry late, I've tired using raspbian lite and it doesn't seem to work. Which is no problem like you said for now i'll use your release files and after then i'll try your scripts with the full version of raspbian after that look into raspbian lite.

The only reason is we've designed a board for the PI Compute Module 3 which we're going to release (open hardware, Altium files) and the aim is to create an image with Qt and some auto configuration features to make it easy to setup + deploy. Was late as I have been testing the board atm.

I'll try using Raspbian full version which will then allow me to test our cm3 motherboard with Qt :D

**Comment by motters** [comment]
I've tested your release files and they work perfectly.

Although i have followed your steps to build qt with your scripts and the EGL qt config issue not long exists but there are build errors. Note, i also had to keep the "sudo sed -i '/./s/^/#/g' $SYSROOT/etc/ld.so.preload" for stage 2 to install the libs.

I've attach some log files for the config, make and install, can you see whats wrong?

There seems to still be an issue with EGL in the make log (search for Error 2), XCB and Sqlite.

[configure.log](https://github.com/GuillaumeLazar/qtrb/files/1414855/configure.log)
[make.log](https://github.com/GuillaumeLazar/qtrb/files/1414856/make.log)
[make_install.log](https://github.com/GuillaumeLazar/qtrb/files/1414857/make_install.log)

**Comment by GuillaumeLazar** [comment]
I'm glad that my release files are working !

I checked your log, the configure log seems good. I have no idea why your build failed. You could maybe try to build Qt with my sysroot image. Depending the result, we will be able to search an error on the sysroot creation step or the build process.

**Comment by motters** [comment]
Good plan. 

I will try and build qt now using your sysroot img and report back.

**Comment by motters** [comment]
I've compiled Qt using your sysroot img and everything seems to work perfectly. I've not tested stage 5 onwards but stage 1-4 (building of qt) works fine.

So obviously the problem seem to be my sysroot img. Can you think of any additional steps that need to take place on the pi before backing up the SD card to an image?

I've uploaded the logs for the build that has just been successful

[configure.log](https://github.com/GuillaumeLazar/qtrb/files/1415152/configure.log)
[make.log](https://github.com/GuillaumeLazar/qtrb/files/1415153/make.log)
[make_install.log](https://github.com/GuillaumeLazar/qtrb/files/1415154/make_install.log)

**Comment by GuillaumeLazar** [comment]
Interesting... As the sysroot creation requires some "manual" action, it seems logical that some errors could appear during its creation.

Have you follow all the steps described in my [comment](https://github.com/GuillaumeLazar/qtrb/issues/1#issuecomment-337187915) (raspbian version, boot the system, expand, ...)? I have some doubts for the "sudo rpi-update". You should try to not do it.

**Comment by motters** [comment]
Yes i followed your instructions as above. 

I used the 2017-09-07-raspbian-stretch version; expanded the file system and enabled ssh via raspi-config. Then i ran "sudo apt update" and then "sudo rpi-update". 

I shutdown the pi and then created an img on Ubuntu by running the below

```bash
sudo dd if=/dev/sdd of=~/qtrb.img
```

Should i re create the sysroot img but this time do not do the "sudo rpi-update" command?

**Comment by GuillaumeLazar** [comment]
Yes a sysroot without the _rpi-update_ could be an interesting test to do.

Note that you can use the **bs** option with 1M or 4M to speed up the copy (in both way). e.g:
`sudo dd if=/dev/sdd of=~/qtrb.img bs=4M`

**Comment by motters** [comment]
I've tried the whole process again but this time without running "rpi-update". And during stage 3 (configuring qt build) there are error about opengl. 

I've uploaded the logs

[configure.log](https://github.com/GuillaumeLazar/qtrb/files/1421208/configure.log)
[make.log](https://github.com/GuillaumeLazar/qtrb/files/1421209/make.log)
[make_install.log](https://github.com/GuillaumeLazar/qtrb/files/1421210/make_install.log)

**Comment by GuillaumeLazar** [comment]
Good news! 

The issue #2 reminds me an important manual step to be able to build Qt. Follow the workflow of [my comment ](https://github.com/GuillaumeLazar/qtrb/issues/1#issuecomment-337187915). 

After you mounted the sysroot on your host computer you must use a script to fix the symlinks. It's explained in the documentation https://wiki.qt.io/RaspberryPi2EGLFS .You should download the script [sysroot-relativelinks.py](https://raw.githubusercontent.com/riscv/riscv-poky/master/scripts/sysroot-relativelinks.py), then you can run it on your mounted sysroot directory.

```bash
wget https://raw.githubusercontent.com/riscv/riscv-poky/master/scripts/sysroot-relativelinks.py
chmod u+x sysroot-relativelinks.py
./sysroot-relativelinks.py sysroot
```

Now you can do the configure, build , deploy stuff :)

**Comment by motters** [comment]
Are sounds like that would solve this issue.

I shall give it another go tomorrow (including the rpi-update) :)

**Comment by motters** [comment]
I ran through the process yesterday and Qt seems to have compiled successfully. I'll test the compiled libraries on Monday however to make sure everything has actually worked.

I did need to [edit](https://github.com/motters/qtrb/blob/master/sysroot-relativelinks.py) the sysroot-relativelinks.py script though; so that it only changed the symlinks in usr/include, opt/vc, usr/lib, and lib. As the current script tried to changed symlinks in directories like proc which resulted in the script failing half way through.
